apis/apps/v1/componentdefinition_types.go

```diff
@@ -0,0 +1,1973 @@
+// across the cluster.
+// The available strategies are `Serial`, `BestEffortParallel`, and `Parallel`.
+//
+// +enum
+// +kubebuilder:validation:Enum={Serial,BestEffortParallel,Parallel}
+type UpdateStrategy string
+
+const (
+	// SerialStrategy indicates that updates are applied one at a time in a sequential manner.
+	// The operator waits for each replica to be updated and ready before proceeding to the next one.
+	// This ensures that only one replica is unavailable at a time during the update process.
+	SerialStrategy UpdateStrategy = "Serial"
+
+	// ParallelStrategy indicates that updates are applied simultaneously to all Pods of a Component.
+	// The replicas are updated in parallel, with the operator updating all replicas concurrently.
+	// This strategy provides the fastest update time but may lead to a period of reduced availability or
+	// capacity during the update process.
+	ParallelStrategy UpdateStrategy = "Parallel"
+
+	// BestEffortParallelStrategy indicates that the replicas are updated in parallel, with the operator making
+	// a best-effort attempt to update as many replicas as possible concurrently
+	// while maintaining the component's availability.
+	// Unlike the `Parallel` strategy, the `BestEffortParallel` strategy aims to ensure that a minimum number
+	// of replicas remain available during the update process to maintain the component's quorum and functionality.
+	//
+	// For example, consider a component with 5 replicas. To maintain the component's availability and quorum,
+	// the operator may allow a maximum of 2 replicas to be simultaneously updated. This ensures that at least
+	// 3 replicas (a quorum) remain available and functional during the update process.
+	//
+	// The `BestEffortParallel` strategy strikes a balance between update speed and component availability.
+	BestEffortParallelStrategy UpdateStrategy = "BestEffortParallel"
+)
+
+// ReplicaRole represents a role that can be assumed by a component instance.
+type ReplicaRole struct {
+	// Defines the role's identifier. It is used to set the "apps.kubeblocks.io/role" label value
+	// on the corresponding object.
+	//
+	// This field is immutable once set.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=32
+	// +kubebuilder:validation:Pattern=`^.*[^\s]+.*$`
+	Name string `json:"name"`
+
+	// Indicates whether a replica assigned this role is capable of providing services.
+	//
+	// This field is immutable once set.
+	//
+	// +kubebuilder:default=false
+	// +optional
+	Serviceable bool `json:"serviceable,omitempty"`
+
+	// Determines if a replica in this role has the authority to perform write operations.
+	// A writable replica can modify data, handle update operations.
+	//
+	// This field is immutable once set.
+	//
+	// +kubebuilder:default=false
+	// +optional
+	Writable bool `json:"writable,omitempty"`
+
```

Comment:
could add more fields to describe a role's abilities or properties, like:

- `Unique bool`, Indicates if there could be only one pod that has such role in the component.
- `Required bool `, Indicates whether at least one replica with this role is required for the component to be considered running.
- `UpdatePriority int`, Describes the priority for updating pods with this role. When updating, all pods are sorted by this priority and updated accordingly.

Reply:
In this way, there is no need to inference which role is leader in controller's code.

Reply:
One more field:

- `SwitchoverBeforeUpdate bool`, // Indicates if a switchover operation is needed before updating.